### PR TITLE
Implement a reversed data transformer for events

### PR DESCRIPTION
**Pull request by integraledelebesgue**
Closes #

## Introduced changes

-

## Checklist

- [ ] Linked relevant issue
- [ ] Updated relevant documentation
- [ ] Added relevant tests
- [ ] Performed self-review of the code
- [ ] Added changes to `CHANGELOG.md`

### CODE REVIEW

**Comment by integraledelebesgue** [comment]
This change is part of the following stack:

- #4299
    - #4300 ◀
        - #4301

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: crates/data-transformer/src/reverse_transformer/event.rs [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+    // Select only top-level events, i.e. such that are not referenced by other events.
+    abi.iter().filter_map(move |entry| match entry {
+        AbiEntry::Event(AbiEvent::Typed(typed_event))
+            if !all_types_in_events.contains(&typed_event_name(typed_event)) =>
+        {
+            Some(typed_event)
+        }
+        _ => None,
+    })
+}
```
Comment (franciszekjob) [inline crates/data-transformer/src/reverse_transformer/event.rs]: nit: Could we use `HashSet` instead of a `Vec` for `all_types_in_events`? Currently, calling `.contains()` inside the `.filter_map` closure gives us `O(N^2)` complexity, while using `HashSet`, lookup will decrease to `O(1)`.

File: crates/data-transformer/src/reverse_transformer/event.rs [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+                .transform_from_keys(&field.r#type, db)
+                .map_err(Into::into),
+            EventFieldKind::Data => self
+                .transform_from_data(&field.r#type, db)
+                .map_err(Into::into),
+            EventFieldKind::Nested => {
+                self.transform_event_type(&field.r#type, SelectorSource::Consume, db)
+            }
+            EventFieldKind::Flat => {
+                self.transform_event_type(&field.r#type, SelectorSource::None, db)
```
Comment (franciszekjob) [inline crates/data-transformer/src/reverse_transformer/event.rs:220]: Could this still fail for a valid flattened struct field, e.g. `OuterEvent { #[flat] inner: InnerEvent, tail: felt252 }`? In that case `InnerEvent` would not have its own selector in the emitted payload, but here we pass `SelectorSource::None` and the struct path still seems to require one.

File: crates/data-transformer/src/reverse_transformer/event.rs [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+
+        for variant in &abi_event_enum.variants {
+            match variant.kind {
+                EventFieldKind::Flat => {
+                    if self.event_type_matches_selector(&variant.r#type, selector)? {
+                        return self.transform_event_type(
+                            &variant.r#type,
+                            SelectorSource::Provided(selector),
+                            db,
+                        );
```
Comment (franciszekjob) [inline crates/data-transformer/src/reverse_transformer/event.rs]: Could this be lossy for `#[flat]` enum variants? It looks like we return only the inner decoded event here, without preserving the outer variant wrapper, which might become ambiguous if multiple flat branches match the same selector.

File: crates/data-transformer/tests/integration/event_reverse_transformer.rs
Comment (franciszekjob) [inline crates/data-transformer/tests/integration/event_reverse_transformer.rs:1]: Could we add a regression test for a directly-emitted typed event that is also referenced as a nested payload in another event? That seems like an edge case behind the top-level filtering logic here.

File: crates/data-transformer/src/reverse_transformer/event.rs [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+}
+
+impl From<ReverseTransformEventError> for InternalEventError {
+    fn from(value: ReverseTransformEventError) -> Self {
+        Self::Other(value)
+    }
+}
+
+/// Transforms a set of event keys and data into a Cairo-like string representation of the event.
+pub fn reverse_transform_event(
```
Comment (franciszekjob) [inline crates/data-transformer/src/reverse_transformer/event.rs:55]: nit: Could we update the ordering so struct/enums are at the top of file?

File: crates/data-transformer/src/reverse_transformer/event.rs [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+#[derive(Debug, thiserror::Error)]
+pub enum ReverseTransformEventError {
+    #[error(r#"Typed event matching emitted selector "{0:#x}" not found in ABI of the contract"#)]
+    EventNotFound(Felt),
+    #[error("untyped event `{0}` is unsupported by reverse transformer")]
+    UnsupportedUntypedEvent(String),
+    #[error("event `{0}` is unsupported by reverse transformer")]
+    UnsupportedEvent(String),
+    #[error("abi is invalid")]
+    InvalidAbi,
```
Comment (MKowalski8) [inline crates/data-transformer/src/reverse_transformer/event.rs]: We could potentailly split it, to sth like:
``` rust
    #[error("abi is invalid: event `{0}` matched selector but left {1} key(s) and {2} data felt(s) unconsumed")]
    InvalidAbi(String, usize, usize),
    #[error("abi is invalid: type `{0}` referenced in event but not found in abi")]
    InvalidAbiMissingType(String),
```

Then in `reverse_transform_event` we will use

```rust
            Ok(decoded) => {
                if first_error.is_none() {
                    first_error = Some(ReverseTransformEventError::InvalidAbi(
                        decoded.to_string(),
                        transformer.keys.len(),
                        transformer.data.len(),
                    ));
                }
            }
```

And in other 3 occurences `InvalidAbiMissingType`

However I'm not sure if this split is really necessary, since with proper ABI from compiler it shouldn't occur. But mayby more detailed error handling won't be so bad after all (and `reverse_data_transformer` don't have the best one). 

Left to consideration

File: crates/data-transformer/src/reverse_transformer/event.rs (hunk 1 above)
Comment (integraledelebesgue) [reply crates/data-transformer/src/reverse_transformer/event.rs]: Done

File: crates/data-transformer/tests/integration/event_reverse_transformer.rs
Comment (integraledelebesgue) [reply crates/data-transformer/tests/integration/event_reverse_transformer.rs:1]: Done, I added a test to verify that the event referenced in an other event can be decoded when emitted directly - https://github.com/foundry-rs/starknet-foundry/blob/fbf382a38edd94ee97e4726390f9695ce7b60f3c/crates/data-transformer/tests/integration/event_reverse_transformer.rs#L152

File: crates/data-transformer/src/reverse_transformer/event.rs (hunk 4 above)
Comment (integraledelebesgue) [reply crates/data-transformer/src/reverse_transformer/event.rs:55]: Done

File: crates/data-transformer/src/reverse_transformer/event.rs (hunk 2 above)
Comment (integraledelebesgue) [reply crates/data-transformer/src/reverse_transformer/event.rs:220]: The struct path interprets the `SelectorSource::None` exactly as an information that the flattened content is being decoded. I added a test: https://github.com/foundry-rs/starknet-foundry/blob/fbf382a38edd94ee97e4726390f9695ce7b60f3c/crates/data-transformer/tests/integration/event_reverse_transformer.rs#L98

File: crates/data-transformer/src/reverse_transformer/event.rs (hunk 5 above)
Comment (integraledelebesgue) [reply crates/data-transformer/src/reverse_transformer/event.rs]: That's a good idea - it's better to have explicit, descriptive errors. I divided the existing error according to your suggestion

**Review by franciszekjob — APPROVED** [review]

**Review by MKowalski8 — APPROVED** [review]
